File: vms-mobile/src/screens/driver/DriverHomeScreen.js
import React, { useState, useCallback, useRef, useEffect } from 'react';
import {
  View, Text, FlatList, TouchableOpacity, Alert,
  StyleSheet, RefreshControl, Platform, Switch,
} from 'react-native';
import * as Location from 'expo-location';
import { useFocusEffect } from '@react-navigation/native';
import { driverApi } from '../../api/services';
import { useAuthStore } from '../../stores/authStore';
import { Card, StatusBadge, GradientHeader, Empty } from '../../components';
import { C } from '../../utils/colors';
import dayjs from 'dayjs';

const LOCATION_INTERVAL_MS = 15_000; // send GPS every 15 seconds while tracking is on

export default function DriverHomeScreen({ navigation }) {
  const [dispatches,   setDispatches]   = useState([]);
  const [profile,      setProfile]      = useState(null);
  const [loading,      setLoading]      = useState(true);
  const [refreshing,   setRefreshing]   = useState(false);
  const [tracking,     setTracking]     = useState(false);
  const [lastGeoEvent, setLastGeoEvent] = useState(null);
  const trackingRef = useRef(false);
  const intervalRef = useRef(null);
  const user = useAuthStore(s => s.user);

  // ── Data load ────────────────────────────────────────────────────────
  async function load() {
    try {
      const [discRes, profRes] = await Promise.all([
        driverApi.todayDispatches(),
        driverApi.me(),
      ]);
      setDispatches(discRes.data.data ?? []);
      setProfile(profRes.data.data);
    } catch { }
    setLoading(false);
    setRefreshing(false);
  }

  useFocusEffect(useCallback(() => { setLoading(true); load(); }, []));

  // ── GPS tracking loop ────────────────────────────────────────────────
  async function startTracking() {
    const { status } = await Location.requestForegroundPermissionsAsync();
    if (status !== 'granted') {
      Alert.alert('Permission needed', 'Location access is required to auto-track your trips.');
      return;
    }
    trackingRef.current = true;
    setTracking(true);

    async function sendLocation() {
      if (!trackingRef.current) return;
      try {
        const loc = await Location.getCurrentPositionAsync({ accuracy: Location.Accuracy.High });
        const res = await driverApi.updateLocation({
          lat:      loc.coords.latitude,
          lng:      loc.coords.longitude,
          speed:    loc.coords.speed != null ? loc.coords.speed * 3.6 : null, // m/s → km/h
          heading:  loc.coords.heading,
          accuracy: loc.coords.accuracy,
        });
        const event = res.data.data;
        if (event?.geofenceTriggered) {
          setLastGeoEvent(event);
          Alert.alert('Geofence Alert 📍', event.message ?? 'Trip status changed');
          load(); // refresh trip list
        }
      } catch { /* silently retry */ }
    }

    await sendLocation();
    intervalRef.current = setInterval(sendLocation, LOCATION_INTERVAL_MS);
  }

  function stopTracking() {
    trackingRef.current = false;
    setTracking(false);
    if (intervalRef.current) { clearInterval(intervalRef.current); intervalRef.current = null; }
  }

  useEffect(() => { return () => stopTracking(); }, []);

  async function toggleTracking(val) {
    if (val) await startTracking();
    else stopTracking();
  }

  // ── Manual start / end ────────────────────────────────────────────────
  async function manualStart(d) {
    Alert.alert('Start Trip', `Start dispatch ${d.dispatchNo}?`, [
      { text: 'Cancel' },
      { text: 'Start', onPress: async () => {
        try { await driverApi.startDispatch(d.id); load(); }
        catch (e) { Alert.alert('Error', e.response?.data?.message ?? 'Failed'); }
      }},
    ]);
  }

  async function manualEnd(d) {
    Alert.alert('End Trip', `Mark dispatch ${d.dispatchNo} as completed?`, [
      { text: 'Cancel' },
      { text: 'End Trip', style: 'destructive', onPress: async () => {
        try { await driverApi.endDispatch(d.id, {}); load(); }
        catch (e) { Alert.alert('Error', e.response?.data?.message ?? 'Failed'); }
      }},
    ]);
  }

  // ── Render ─────────────────────────────────────────────────────────────
  function renderDispatch({ item: d }) {
    const isActive = d.status === 'in_progress';
    const canStart = d.status === 'approved' || d.status === 'pending';
    const border   = isActive ? C.primary : C.statusColor(d.status);

    return (
      <TouchableOpacity
        activeOpacity={0.88}
        onPress={() => navigation.navigate('DispatchDetail', { id: d.id })}
        style={[s.card, { borderLeftColor: border, borderLeftWidth: 4 }]}
      >
        <View style={{ flexDirection: 'row', justifyContent: 'space-between', alignItems: 'flex-start', marginBottom: 6 }}>
          <View>
            <Text style={s.dispatchNo}>{d.dispatchNo}</Text>
            {d.date && <Text style={s.dateText}>{dayjs(d.date).format('DD MMM YYYY')}</Text>}
          </View>
          <StatusBadge status={d.status} />
        </View>

        <View style={{ gap: 4 }}>
          <View style={{ flexDirection: 'row', alignItems: 'center', gap: 8 }}>
            <View style={s.dotA} />
            <Text style={[s.locText, { color: C.primary }]} numberOfLines={1}>{d.origin}</Text>
          </View>
          <View style={{ flexDirection: 'row', alignItems: 'center', gap: 8 }}>
            <View style={s.dotB} />
            <Text style={[s.locText, { color: C.emerald }]} numberOfLines={1}>{d.destination}</Text>
          </View>
        </View>

        {d.purpose && <Text style={s.purpose} numberOfLines={1}>{d.purpose}</Text>}

        <View style={s.actions}>
          {canStart && (
            <TouchableOpacity style={[s.actionBtn, { backgroundColor: C.emerald }]} onPress={() => manualStart(d)}>
              <Text style={s.actionBtnText}>▶ Start Trip</Text>
            </TouchableOpacity>
          )}
          {isActive && (
            <>
              <TouchableOpacity style={[s.actionBtn, { backgroundColor: C.primary }]}
                onPress={() => navigation.navigate('ActiveTrip', { id: d.id })}>
                <Text style={s.actionBtnText}>🗺 Live Map</Text>
              </TouchableOpacity>
              <TouchableOpacity style={[s.actionBtn, { backgroundColor: C.rose }]} onPress={() => manualEnd(d)}>
                <Text style={s.actionBtnText}>■ End Trip</Text>
              </TouchableOpacity>
            </>
          )}
        </View>
      </TouchableOpacity>
    );
  }

  const todaySummary = profile?.tripStats;

  return (
    <View style={{ flex: 1, backgroundColor: C.bg }}>
      <GradientHeader
        title={`Hello, ${user?.fullName?.split(' ')[0] ?? 'Driver'} 👋`}
        subtitle="Today's trip schedule"
        color1="#0891b2" color2="#6366f1"
      />

      {/* GPS tracking toggle */}
      <View style={s.trackingBar}>
        <View>
          <Text style={s.trackingTitle}>{tracking ? '🟢 GPS Tracking Active' : '⚪ GPS Tracking Off'}</Text>
          <Text style={s.trackingDesc}>{tracking ? 'Auto-starts/ends trips based on location' : 'Turn on to enable geofence auto-detection'}</Text>
        </View>
        <Switch
          value={tracking}
          onValueChange={toggleTracking}
          trackColor={{ true: C.emerald, false: C.border }}
          thumbColor={tracking ? '#fff' : C.textMuted}
        />
      </View>

      {/* Last geofence event */}
      {lastGeoEvent?.geofenceTriggered && (
        <View style={[s.eventBanner, { backgroundColor: lastGeoEvent.geofenceEvent === 'TRIP_COMPLETED' ? C.emeraldBg : C.primaryBg }]}>
          <Text style={{ fontSize: 13, fontWeight: '700', color: lastGeoEvent.geofenceEvent === 'TRIP_COMPLETED' ? C.emerald : C.primary }}>
            {lastGeoEvent.geofenceEvent === 'TRIP_STARTED' ? '▶ Trip auto-started' :
             lastGeoEvent.geofenceEvent === 'TRIP_COMPLETED' ? '✓ Trip auto-completed' :
             '📍 Approaching destination'}
          </Text>
          <Text style={{ fontSize: 12, color: C.textMuted, marginTop: 2 }}>{lastGeoEvent.message}</Text>
        </View>
      )}

      {/* Stats row */}
      {todaySummary && (
        <View style={s.statsRow}>
          {[
            { label: 'Approved', value: todaySummary.approved, color: C.emerald },
            { label: 'Active',   value: todaySummary.inProgress, color: C.primary },
            { label: 'Done',     value: todaySummary.completed, color: C.textMuted },
          ].map(st => (
            <View key={st.label} style={s.statCell}>
              <Text style={[s.statValue, { color: st.color }]}>{st.value}</Text>
              <Text style={s.statLabel}>{st.label}</Text>
            </View>
          ))}
        </View>
      )}

      <Text style={s.sectionLabel}>Today's Dispatches</Text>

      <FlatList
        data={dispatches}
        keyExtractor={i => String(i.id)}
        renderItem={renderDispatch}
        contentContainerStyle={{ paddingHorizontal: 16, paddingBottom: 40 }}
        refreshControl={<RefreshControl refreshing={refreshing} onRefresh={() => { setRefreshing(true); load(); }} />}
        ListEmptyComponent={!loading && (
          <Empty icon="🚗" title="No trips today" subtitle="Your approved trips will appear here" />
        )}
      />
    </View>
  );
}

const s = StyleSheet.create({
  trackingBar:  { flexDirection: 'row', alignItems: 'center', justifyContent: 'space-between',
    backgroundColor: C.white, padding: 14, paddingHorizontal: 18,
    borderBottomWidth: 1, borderBottomColor: C.borderLight },
  trackingTitle:{ fontSize: 14, fontWeight: '700', color: C.text },
  trackingDesc: { fontSize: 11, color: C.textMuted, marginTop: 1 },
  eventBanner:  { margin: 12, borderRadius: 12, padding: 12,
    borderWidth: 1, borderColor: C.borderLight },
  statsRow:     { flexDirection: 'row', backgroundColor: C.white,
    borderBottomWidth: 1, borderBottomColor: C.borderLight },
  statCell:     { flex: 1, alignItems: 'center', paddingVertical: 12 },
  statValue:    { fontSize: 22, fontWeight: '900' },
  statLabel:    { fontSize: 11, color: C.textMuted, marginTop: 2, fontWeight: '600' },
  sectionLabel: { fontSize: 12, fontWeight: '800', color: C.textMuted, textTransform: 'uppercase',
    letterSpacing: 0.8, paddingHorizontal: 18, paddingTop: 14, paddingBottom: 6 },
  card:         { backgroundColor: C.white, borderRadius: 14, padding: 16, marginBottom: 10,
    borderWidth: 1, borderColor: C.borderLight,
    shadowColor: '#000', shadowOpacity: 0.04, shadowOffset: { width: 0, height: 2 }, shadowRadius: 8, elevation: 2 },
  dispatchNo:   { fontWeight: '800', fontSize: 15, color: C.text },
  dateText:     { fontSize: 12, color: C.textMuted, marginTop: 1 },
  dotA:         { width: 9, height: 9, borderRadius: 5, backgroundColor: C.primary },
  dotB:         { width: 9, height: 9, borderRadius: 2, backgroundColor: C.emerald },
  locText:      { fontSize: 13, flex: 1, fontWeight: '500' },
  purpose:      { fontSize: 12, color: C.textMuted, marginTop: 8 },
  actions:      { flexDirection: 'row', gap: 8, marginTop: 12 },
  actionBtn:    { flex: 1, paddingVertical: 9, borderRadius: 10, alignItems: 'center' },
  actionBtnText:{ color: '#fff', fontWeight: '700', fontSize: 13 },
});

File: vms_flutter/lib/screens/driver/driver_home_screen.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:geolocator/geolocator.dart';
import 'package:intl/intl.dart';
import '../../core/api.dart';
import '../../core/constants.dart';
import '../../models/models.dart';
import '../../widgets/widgets.dart';
import 'active_trip_screen.dart';
import 'dispatch_detail_screen.dart';

class DriverHomeScreen extends StatefulWidget {
  const DriverHomeScreen({super.key});
  @override State<DriverHomeScreen> createState() => _DriverHomeScreenState();
}

class _DriverHomeScreenState extends State<DriverHomeScreen> {
  List<Dispatch> _dispatches = [];
  bool           _loading    = true;
  bool           _tracking   = false;
  String?        _lastEvent;
  Timer?         _timer;

  @override void initState() { super.initState(); _load(); }
  @override void dispose()   { _stopTracking(); super.dispose(); }

  Future<void> _load() async {
    setState(() => _loading = true);
    try {
      final res = await ApiClient.get('/driver/dispatches/today');
      setState(() => _dispatches = (res['data'] as List? ?? [])
        .map((j) => Dispatch.fromJson(j as Map<String,dynamic>)).toList());
    } catch (_) {}
    setState(() => _loading = false);
  }

  // ── GPS Tracking ─────────────────────────────────────────────────────────────
  Future<void> _toggleTracking(bool on) async {
    if (on) {
      final perm = await Geolocator.requestPermission();
      if (perm == LocationPermission.denied || perm == LocationPermission.deniedForever) {
        if (mounted) ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(content: Text('Location permission required')));
        return;
      }
      setState(() => _tracking = true);
      _sendLocation(); // immediate
      _timer = Timer.periodic(const Duration(seconds: 15), (_) => _sendLocation());
    } else {
      _stopTracking();
    }
  }

  void _stopTracking() { _timer?.cancel(); setState(() => _tracking = false); }

  Future<void> _sendLocation() async {
    try {
      final pos = await Geolocator.getCurrentPosition(
        desiredAccuracy: LocationAccuracy.high);
      final res = await ApiClient.post('/driver/location', {
        'lat':      pos.latitude,
        'lng':      pos.longitude,
        'speed':    pos.speed * 3.6,  // m/s → km/h
        'heading':  pos.heading,
        'accuracy': pos.accuracy,
      });
      final data  = res['data'] as Map<String,dynamic>? ?? {};
      final event = data['geofenceEvent'] as String?;
      if (event != null && mounted) {
        setState(() => _lastEvent = data['message'] as String?);
        ScaffoldMessenger.of(context).showSnackBar(SnackBar(
          content: Text(data['message'] as String? ?? 'Geofence triggered'),
          backgroundColor: event == 'TRIP_COMPLETED' ? AppColors.emerald : AppColors.primary,
          duration: const Duration(seconds: 4)));
        _load();
      }
    } catch (_) {}
  }

  // ── Manual start / end ─────────────────────────────────────────────────────
  Future<void> _startDispatch(Dispatch d) async {
    final ok = await showDialog<bool>(context: context,
      builder: (_) => AlertDialog(title: const Text('Start Trip'),
        content: Text('Start dispatch ${d.dispatchNo}?'),
        actions: [
          TextButton(onPressed: () => Navigator.pop(context, false), child: const Text('Cancel')),
          TextButton(onPressed: () => Navigator.pop(context, true),
            child: const Text('Start', style: TextStyle(color: AppColors.emerald))),
        ]));
    if (ok != true) return;
    try {
      await ApiClient.post('/driver/dispatches/${d.id}/start', {});
      _load();
    } catch (e) {
      if (mounted) ScaffoldMessenger.of(context)
        .showSnackBar(SnackBar(content: Text(e.toString())));
    }
  }

  Future<void> _endDispatch(Dispatch d) async {
    final ok = await showDialog<bool>(context: context,
      builder: (_) => AlertDialog(title: const Text('End Trip'),
        content: Text('Mark ${d.dispatchNo} as completed?'),
        actions: [
          TextButton(onPressed: () => Navigator.pop(context, false), child: const Text('Cancel')),
          TextButton(onPressed: () => Navigator.pop(context, true),
            child: const Text('End Trip', style: TextStyle(color: AppColors.rose))),
        ]));
    if (ok != true) return;
    try {
      await ApiClient.post('/driver/dispatches/${d.id}/end', {});
      _load();
    } catch (e) {
      if (mounted) ScaffoldMessenger.of(context)
        .showSnackBar(SnackBar(content: Text(e.toString())));
    }
  }

  @override
  Widget build(BuildContext context) => Scaffold(
    backgroundColor: AppColors.bg,
    appBar: GradientHeader(
      title: "Today's Trips",
      subtitle: '${_dispatches.length} dispatches',
      color1: const Color(0xFF0891B2), color2: AppColors.primary,
      trailing: IconButton(icon: const Icon(Icons.refresh, color: Colors.white),
        onPressed: _load),
    ),
    body: Column(children: [
      // GPS toggle bar
      Container(color: AppColors.white, padding: const EdgeInsets.symmetric(
        horizontal: 18, vertical: 12),
        child: Row(children: [
          Expanded(child: Column(crossAxisAlignment: CrossAxisAlignment.start, children: [
            Text(_tracking ? '🟢 GPS Tracking Active' : '⚪ GPS Tracking Off',
              style: const TextStyle(fontWeight: FontWeight.w700, fontSize: 14)),
            Text(_tracking ? 'Auto-starts/ends trips via geofence'
              : 'Enable to auto-detect trip start/end',
              style: const TextStyle(fontSize: 12, color: AppColors.textMuted)),
          ])),
          Switch(value: _tracking, onChanged: _toggleTracking,
            activeColor: AppColors.emerald),
        ]),
      ),

      // Geofence event banner
      if (_lastEvent != null) Container(
        margin: const EdgeInsets.symmetric(horizontal: 14, vertical: 8),
        padding: const EdgeInsets.all(12),
        decoration: BoxDecoration(color: AppColors.primaryBg,
          borderRadius: BorderRadius.circular(12),
          border: Border.all(color: const Color(0x406366F1))),
        child: Row(children: [
          const Icon(Icons.location_on, color: AppColors.primary, size: 18),
          const SizedBox(width: 8),
          Expanded(child: Text(_lastEvent!,
            style: const TextStyle(color: AppColors.primary, fontWeight: FontWeight.w600))),
        ]),
      ),

      // Trip list
      Expanded(child: _loading
        ? const Center(child: CircularProgressIndicator())
        : _dispatches.isEmpty
        ? const EmptyState(emoji: '🚗', title: "No trips today",
            subtitle: 'Your approved trips will appear here')
        : RefreshIndicator(onRefresh: _load,
            child: ListView.builder(
              padding: const EdgeInsets.all(14),
              itemCount: _dispatches.length,
              itemBuilder: (_, i) => _buildCard(_dispatches[i])))),
    ]),
  );

  Widget _buildCard(Dispatch d) {
    final isActive = d.status == 'in_progress';
    final canStart = d.status == 'approved' || d.status == 'pending';
    return GestureDetector(
      onTap: () async {
        await Navigator.push(context, MaterialPageRoute(
          builder: (_) => DispatchDetailScreen(id: d.id)));
        _load();
      },
      child: Container(
        margin: const EdgeInsets.only(bottom: 12),
        decoration: BoxDecoration(
          color: AppColors.white, borderRadius: BorderRadius.circular(16),
          border: Border.all(color: const Color(0x1F6366F1)),
          borderLeft: Border(left: BorderSide(
            color: AppColors.statusColor(d.status), width: 4))),
        child: Padding(padding: const EdgeInsets.all(16), child: Column(
          crossAxisAlignment: CrossAxisAlignment.start, children: [
          Row(mainAxisAlignment: MainAxisAlignment.spaceBetween, children: [
            Column(crossAxisAlignment: CrossAxisAlignment.start, children: [
              Text(d.dispatchNo, style: const TextStyle(fontWeight: FontWeight.w800,
                fontSize: 15, color: AppColors.text)),
              if (d.date != null) Text(DateFormat('dd MMM yyyy')
                .format(DateTime.parse(d.date!)),
                style: const TextStyle(fontSize: 12, color: AppColors.textMuted)),
            ]),
            StatusBadge(d.status),
          ]),
          const SizedBox(height: 10),
          RouteConnector(from: d.origin, to: d.destination),
          if (d.purpose?.isNotEmpty == true) ...[
            const SizedBox(height: 6),
            Text(d.purpose!, style: const TextStyle(fontSize: 12, color: AppColors.textMuted),
              maxLines: 1, overflow: TextOverflow.ellipsis),
          ],
          const SizedBox(height: 12),
          Row(children: [
            if (canStart) Expanded(child: _btn('▶  Start Trip', AppColors.emerald,
              () => _startDispatch(d))),
            if (isActive) ...[
              Expanded(child: _btn('🗺  Live Map', AppColors.primary,
                () => Navigator.push(context, MaterialPageRoute(
                  builder: (_) => ActiveTripScreen(dispatch: d))))),
              const SizedBox(width: 8),
              Expanded(child: _btn('■  End Trip', AppColors.rose,
                () => _endDispatch(d))),
            ],
          ]),
        ])),
      ),
    );
  }

  Widget _btn(String label, Color color, VoidCallback onTap) =>
    GestureDetector(onTap: onTap,
      child: Container(
        padding: const EdgeInsets.symmetric(vertical: 9),
        margin: const EdgeInsets.only(right: 8),
        alignment: Alignment.center,
        decoration: BoxDecoration(color: color, borderRadius: BorderRadius.circular(10)),
        child: Text(label, style: const TextStyle(color: Colors.white,
          fontWeight: FontWeight.w700, fontSize: 13)),
      ));
}

// Custom border decoration
extension _BoxDecEx on BoxDecoration {
  BoxDecoration copyWithBorderLeft(BorderSide side) => BoxDecoration(
    color: color, image: image, border: Border(
      left:   side,
      top:    (border as Border?)?.top    ?? BorderSide.none,
      right:  (border as Border?)?.right  ?? BorderSide.none,
      bottom: (border as Border?)?.bottom ?? BorderSide.none),
    borderRadius: borderRadius, boxShadow: boxShadow, gradient: gradient);
}
